File: examples/mobile-government/ios/Sources/MobileAPI.swift
// SPDX-License-Identifier: Apache-2.0
import EmiliaMobile
import Foundation

private final class NoRedirectSessionDelegate: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

struct MobileAPI: Sendable {
    struct PairingResponse: Decodable, Sendable {
        let accessToken: String
        let expiresAt: String
        let approverID: String
        let profileID: String

        enum CodingKeys: String, CodingKey {
            case accessToken = "access_token"
            case expiresAt = "expires_at"
            case approverID = "approver_id"
            case profileID = "profile_id"
        }
    }

    struct InboxResponse: Decodable, Sendable {
        let approverID: String
        let actions: [InboxAction]

        enum CodingKeys: String, CodingKey {
            case approverID = "approver_id"
            case actions
        }
    }

    struct InboxAction: Decodable, Identifiable, Sendable, Equatable {
        let actionReference: String
        let title: String
        let summary: String
        let risk: String
        let materialFields: [String: String]
        let expiresAt: String
        let createdAt: String

        var id: String { actionReference }

        enum CodingKeys: String, CodingKey {
            case actionReference = "action_reference"
            case title, summary, risk
            case materialFields = "material_fields"
            case expiresAt = "expires_at"
            case createdAt = "created_at"
        }
    }

    struct IssueResponse: Decodable {
        let ok: Bool
        let verdict: String
        let challenge: EmiliaMobileChallenge?
    }

    struct EnrollmentIssueResponse: Decodable {
        let ok: Bool
        let verdict: String
        let challenge: EmiliaMobileEnrollmentChallenge?
    }

    struct VerificationResponse: Decodable {
        let valid: Bool
        let verdict: String
        let decision: String?
        let reason: String?
        let contextHash: String?

        enum CodingKeys: String, CodingKey {
            case valid, verdict, decision, reason
            case contextHash = "context_hash"
        }
    }

    private struct CeremonyRecoveryResponse: Decodable {
        let committed: Bool
        let outcome: String
        let result: VerificationResponse?
    }

    struct EnrollmentResponse: Decodable {
        struct Enrollment: Decodable {
            let deviceKeyID: String
            let attestationKeyID: String

            enum CodingKeys: String, CodingKey {
                case deviceKeyID = "device_key_id"
                case attestationKeyID = "attestation_key_id"
            }
        }
        let ok: Bool
        let verdict: String
        let enrollment: Enrollment?
    }

    let baseURL: URL
    let accessToken: String?
    let session: URLSession
    private static let productionBaseURL = "https://www.emiliaprotocol.ai/api/"
    private static let maximumResponseBytes = 1_048_576

    init(baseURL: URL, accessToken: String? = nil, session: URLSession? = nil) {
        precondition(baseURL.absoluteString == Self.productionBaseURL, "The approval API identity is not pinned")
        self.baseURL = baseURL
        self.accessToken = accessToken
        if let session {
            self.session = session
        } else {
            let configuration = URLSessionConfiguration.ephemeral
            configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
            configuration.urlCache = nil
            configuration.timeoutIntervalForRequest = 15
            configuration.timeoutIntervalForResource = 30
            self.session = URLSession(
                configuration: configuration,
                delegate: NoRedirectSessionDelegate(),
                delegateQueue: nil
            )
        }
    }

    func exchangePairing(code: String, appID: String) async throws -> PairingResponse {
        let body: [String: String] = [
            "pairing_code": code,
            "platform": "ios",
            "app_id": appID,
        ]
        return try await post("v1/mobile/pairings/exchange", body, authenticated: false)
    }

    func inbox() async throws -> [InboxAction] {
        let response: InboxResponse = try await get("v1/mobile/inbox")
        return response.actions
    }

    func issueChallenge(
        requestID: String,
        approverID: String,
        decision: String,
        profileID: String,
        appID: String,
        deviceKeyID: String
    ) async throws -> EmiliaMobileChallenge {
        let body: [String: String] = [
            "profile_id": profileID,
            "action_reference": requestID,
            "approver_id": approverID,
            "decision": decision,
            "platform": "ios",
            "app_id": appID,
            "device_key_id": deviceKeyID,
        ]
        let response: IssueResponse = try await post("v1/mobile/challenges", body)
        guard response.ok, let challenge = response.challenge else { throw APIError.refused(response.verdict) }
        return challenge
    }

    func verify(challenge: EmiliaMobileChallenge, response: EmiliaMobileCeremonyResponse) async throws -> VerificationResponse {
        struct Request: Encodable {
            let challenge: EmiliaMobileChallenge
            let response: EmiliaMobileCeremonyResponse
        }
        guard let expectedDecision = challenge.authorizationContext.decision,
              response.decision == expectedDecision
        else {
            throw APIError.refused("decision_mismatch")
        }
        let expectedContextHash = try EmiliaCanonicalJSON.digest(challenge.authorizationContext)
        do {
            return try await post("v1/mobile/ceremonies", Request(challenge: challenge, response: response))
        } catch APIError.transport {
            return try await recoverCeremonyResult(
                challengeID: challenge.challengeID,
                expectedDecision: challenge.authorizationContext.decision ?? expectedDecision,
                expectedContextHash: expectedContextHash
            )
        }
    }

    func issueEnrollment(approverID: String, appID: String) async throws -> EmiliaMobileEnrollmentChallenge {
        let body: [String: String] = [
            "approver_id": approverID,
            "platform": "ios",
            "app_id": appID,
        ]
        let response: EnrollmentIssueResponse = try await post("v1/mobile/enrollments/challenges", body)
        guard response.ok, let challenge = response.challenge else { throw APIError.refused(response.verdict) }
        return challenge
    }

    func completeEnrollment(
        challenge: EmiliaMobileEnrollmentChallenge,
        response: EmiliaMobileEnrollmentResponse
    ) async throws -> EnrollmentResponse {
        struct Request: Encodable {
            let challenge: EmiliaMobileEnrollmentChallenge
            let response: EmiliaMobileEnrollmentResponse
        }
        return try await post("v1/mobile/enrollments", Request(challenge: challenge, response: response))
    }

    func revokeSession() async throws {
        struct RevocationResponse: Decodable { let revoked: Bool }
        var request = URLRequest(url: endpoint("v1/mobile/session"))
        request.httpMethod = "DELETE"
        authorize(&request)
        let response: RevocationResponse = try await execute(request)
        guard response.revoked else { throw APIError.refused("session_not_revoked") }
    }

    private func get<Response: Decodable>(_ path: String) async throws -> Response {
        var request = URLRequest(url: endpoint(path))
        request.httpMethod = "GET"
        authorize(&request)
        return try await execute(request)
    }

    private func recoverCeremonyResult(
        challengeID: String,
        expectedDecision: String,
        expectedContextHash: String
    ) async throws -> VerificationResponse {
        guard challengeID.range(
            of: #"^[A-Za-z0-9:_.@-]{8,256}$"#,
            options: .regularExpression
        ) != nil else { throw APIError.outcomeUnknown }
        do {
            let recovery: CeremonyRecoveryResponse = try await get("v1/mobile/ceremonies/\(challengeID)")
            guard recovery.committed,
                  recovery.outcome == "committed",
                  let result = recovery.result,
                  result.valid,
                  result.verdict == "verified",
                  result.decision == expectedDecision,
                  result.reason == nil,
                  result.contextHash == expectedContextHash
            else { throw APIError.outcomeUnknown }
            return result
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            throw APIError.outcomeUnknown
        }
    }

    private func post<Request: Encodable, Response: Decodable>(
        _ path: String,
        _ body: Request,
        authenticated: Bool = true
    ) async throws -> Response {
        var request = URLRequest(url: endpoint(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        if authenticated { authorize(&request) }
        return try await execute(request)
    }

    private func endpoint(_ path: String) -> URL {
        path.split(separator: "/").reduce(baseURL) { partial, component in
            partial.appendingPathComponent(String(component))
        }
    }

    private func authorize(_ request: inout URLRequest) {
        if let accessToken, !accessToken.isEmpty {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }
    }

    private func execute<Response: Decodable>(_ request: URLRequest) async throws -> Response {
        do {
            let (bytes, rawResponse) = try await session.bytes(for: request)
            guard let response = rawResponse as? HTTPURLResponse else { throw APIError.transport }
            guard response.url?.scheme == "https",
                  response.url?.host == "www.emiliaprotocol.ai",
                  response.url?.path.hasPrefix("/api/") == true,
                  response.mimeType == "application/json",
                  response.expectedContentLength <= Int64(Self.maximumResponseBytes)
            else { throw APIError.transport }
            var data = Data()
            data.reserveCapacity(max(0, min(Int(response.expectedContentLength), Self.maximumResponseBytes)))
            for try await byte in bytes {
                guard data.count < Self.maximumResponseBytes else { throw APIError.transport }
                data.append(byte)
            }
            guard (200..<300).contains(response.statusCode) else {
                let problem = try? JSONDecoder().decode(ProblemResponse.self, from: data)
                if response.statusCode == 401 { throw APIError.sessionExpired }
                if (500..<600).contains(response.statusCode) { throw APIError.transport }
                throw APIError.refused(problem?.detail ?? problem?.reason ?? "HTTP \(response.statusCode)")
            }
            return try JSONDecoder().decode(Response.self, from: data)
        } catch let error as APIError {
            throw error
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            throw APIError.transport
        }
    }
}

private extension EmiliaJSONValue {
    var decision: String? { objectValue?["decision"]?.stringValue }
}

private struct ProblemResponse: Decodable {
    let detail: String?
    let reason: String?
}

enum APIError: LocalizedError {
    case transport
    case outcomeUnknown
    case sessionExpired
    case refused(String)

    var errorDescription: String? {
        switch self {
        case .transport:
            return "The approval service is unavailable."
        case .outcomeUnknown:
            return "The approval outcome is unknown. Do not retry or assume it was not authorized."
        case .sessionExpired:
            return "This device connection has expired. Pair it again to continue."
        case .refused(let verdict):
            return "The request was refused: \(verdict)"
        }
    }
}

File: sdks/kotlin-mobile/sample/src/main/kotlin/ai/emiliaprotocol/approver/MobileApi.kt
// SPDX-License-Identifier: Apache-2.0
package ai.emiliaprotocol.approver

import ai.emiliaprotocol.mobile.EmiliaMobileCeremonyResponse
import ai.emiliaprotocol.mobile.EmiliaMobileChallenge
import ai.emiliaprotocol.mobile.EmiliaMobileEnrollmentChallenge
import ai.emiliaprotocol.mobile.EmiliaMobileEnrollmentResponse
import java.io.ByteArrayOutputStream
import java.net.URL
import javax.net.ssl.HttpsURLConnection
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

@Serializable
data class PairingResponse(
    @SerialName("access_token") val accessToken: String,
    @SerialName("expires_at") val expiresAt: String,
    @SerialName("approver_id") val approverId: String,
    @SerialName("profile_id") val profileId: String,
)

@Serializable
data class InboxResponse(
    @SerialName("approver_id") val approverId: String,
    val actions: List<InboxAction>,
)

@Serializable
data class InboxAction(
    @SerialName("action_reference") val actionReference: String,
    val title: String,
    val summary: String,
    val risk: String,
    @SerialName("material_fields") val materialFields: JsonObject,
    @SerialName("expires_at") val expiresAt: String,
    @SerialName("created_at") val createdAt: String,
)

@Serializable
private data class ChallengeResponse(
    val ok: Boolean,
    val verdict: String,
    val challenge: EmiliaMobileChallenge? = null,
)

@Serializable
private data class EnrollmentChallengeResponse(
    val ok: Boolean,
    val verdict: String,
    val challenge: EmiliaMobileEnrollmentChallenge? = null,
)

@Serializable
data class EnrollmentRecord(
    @SerialName("device_key_id") val deviceKeyId: String,
    @SerialName("attestation_key_id") val attestationKeyId: String,
)

@Serializable
data class EnrollmentResult(
    val ok: Boolean,
    val verdict: String,
    val enrollment: EnrollmentRecord? = null,
)

@Serializable
data class CeremonyResult(
    val valid: Boolean,
    val verdict: String,
    val decision: String? = null,
    val reason: String? = null,
    @SerialName("context_hash") val contextHash: String? = null,
)

@Serializable
private data class RevocationResult(val revoked: Boolean)

@Serializable
private data class Problem(val detail: String? = null, val reason: String? = null, val verdict: String? = null)

class MobileApi(
    rawBaseUrl: String,
    private val accessToken: String? = null,
) {
    private val json = Json { ignoreUnknownKeys = true; explicitNulls = true }
    private val baseUrl = URL(rawBaseUrl).also {
        require(it.protocol == "https") { "The approval service must use HTTPS" }
        require(it.host.isNotBlank()) { "The approval service host is missing" }
        require(!BuildConfig.PRODUCTION_RELEASE || it.toExternalForm() == PRODUCTION_BASE_URL) {
            "The production approval API identity is not pinned"
        }
    }

    suspend fun exchangePairing(code: String, appId: String): PairingResponse = post(
        "v1/mobile/pairings/exchange",
        buildJsonObject {
            put("pairing_code", code)
            put("platform", "android")
            put("app_id", appId)
        },
        authenticated = false,
    )

    suspend fun inbox(): List<InboxAction> = get<InboxResponse>("v1/mobile/inbox").actions

    suspend fun issueEnrollment(approverId: String, appId: String): EmiliaMobileEnrollmentChallenge {
        val result = post<EnrollmentChallengeResponse>(
            "v1/mobile/enrollments/challenges",
            buildJsonObject {
                put("approver_id", approverId)
                put("platform", "android")
                put("app_id", appId)
            },
        )
        return result.challenge?.takeIf { result.ok } ?: throw MobileApiException.Refused(result.verdict)
    }

    suspend fun completeEnrollment(
        challenge: EmiliaMobileEnrollmentChallenge,
        response: EmiliaMobileEnrollmentResponse,
    ): EnrollmentResult = post(
        "v1/mobile/enrollments",
        buildJsonObject {
            put("challenge", json.encodeToJsonElement(EmiliaMobileEnrollmentChallenge.serializer(), challenge))
            put("response", json.encodeToJsonElement(EmiliaMobileEnrollmentResponse.serializer(), response))
        },
    )

    suspend fun issueChallenge(
        actionReference: String,
        approverId: String,
        decision: String,
        profileId: String,
        appId: String,
        deviceKeyId: String,
    ): EmiliaMobileChallenge {
        val result = post<ChallengeResponse>(
            "v1/mobile/challenges",
            buildJsonObject {
                put("profile_id", profileId)
                put("action_reference", actionReference)
                put("approver_id", approverId)
                put("decision", decision)
                put("platform", "android")
                put("app_id", appId)
                put("device_key_id", deviceKeyId)
            },
        )
        return result.challenge?.takeIf { result.ok } ?: throw MobileApiException.Refused(result.verdict)
    }

    suspend fun verify(
        challenge: EmiliaMobileChallenge,
        response: EmiliaMobileCeremonyResponse,
    ): CeremonyResult = post(
        "v1/mobile/ceremonies",
        buildJsonObject {
            put("challenge", json.encodeToJsonElement(EmiliaMobileChallenge.serializer(), challenge))
            put("response", json.encodeToJsonElement(EmiliaMobileCeremonyResponse.serializer(), response))
        },
    )

    suspend fun revokeSession() {
        if (!delete<RevocationResult>("v1/mobile/session").revoked) {
            throw MobileApiException.Refused("session_not_revoked")
        }
    }

    private suspend inline fun <reified T> get(path: String): T = request("GET", path, null, true)

    private suspend inline fun <reified T> post(
        path: String,
        body: JsonElement,
        authenticated: Boolean = true,
    ): T = request("POST", path, body, authenticated)

    private suspend inline fun <reified T> delete(path: String): T = request("DELETE", path, null, true)

    private suspend inline fun <reified T> request(
        method: String,
        path: String,
        body: JsonElement?,
        authenticated: Boolean,
    ): T = withContext(Dispatchers.IO) {
        val endpoint = URL(baseUrl, path)
        check(endpoint.protocol == "https" && endpoint.host == baseUrl.host) { "Unsafe approval endpoint" }
        val connection = endpoint.openConnection() as HttpsURLConnection
        try {
            connection.requestMethod = method
            connection.instanceFollowRedirects = false
            connection.useCaches = false
            connection.connectTimeout = 10_000
            connection.readTimeout = 20_000
            connection.setRequestProperty("Accept", "application/json")
            if (authenticated) {
                val token = accessToken?.takeIf { it.isNotBlank() } ?: throw MobileApiException.SessionExpired
                connection.setRequestProperty("Authorization", "Bearer $token")
            }
            if (body != null) {
                connection.doOutput = true
                connection.setRequestProperty("Content-Type", "application/json")
                connection.outputStream.use { it.write(body.toString().toByteArray(Charsets.UTF_8)) }
            }
            val status = connection.responseCode
            val contentType = connection.contentType?.substringBefore(';')?.trim()?.lowercase()
            if (connection.url.protocol != "https" || connection.url.host != baseUrl.host
                || contentType != "application/json") throw MobileApiException.Transport
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val bytes = stream?.use { input ->
                val output = ByteArrayOutputStream()
                val chunk = ByteArray(8_192)
                var total = 0
                while (true) {
                    val read = input.read(chunk)
                    if (read < 0) break
                    total += read
                    if (total > MAX_RESPONSE_BYTES) throw MobileApiException.Transport
                    output.write(chunk, 0, read)
                }
                output.toByteArray()
            } ?: ByteArray(0)
            if (status == 401) throw MobileApiException.SessionExpired
            if (status !in 200..299) {
                val problem = runCatching { json.decodeFromString<Problem>(bytes.toString(Charsets.UTF_8)) }.getOrNull()
                throw MobileApiException.Refused(problem?.detail ?: problem?.reason ?: problem?.verdict ?: "HTTP $status")
            }
            json.decodeFromString<T>(bytes.toString(Charsets.UTF_8))
        } catch (error: MobileApiException) {
            throw error
        } catch (_: Exception) {
            throw MobileApiException.Transport
        } finally {
            connection.disconnect()
        }
    }

    private companion object {
        const val MAX_RESPONSE_BYTES = 1_048_576
        const val PRODUCTION_BASE_URL = "https://www.emiliaprotocol.ai/api/"
    }
}

sealed class MobileApiException(message: String) : Exception(message) {
    data object Transport : MobileApiException("The approval service is unavailable. Nothing was authorized.")
    data object SessionExpired : MobileApiException("This device connection has expired.")
    data class Refused(val detail: String) : MobileApiException("The request was refused: $detail")
}
